### PR TITLE
pgwire: optimize writing bytes and strings

### PR DESCRIPTION
**Pull request by ajwerner**
This was a shocking 80% of allocations in YCSB/E and over 12% of CPU.

Epic: None

Release note: None

### CODE REVIEW

**Review by knz — APPROVED** [review]
niiice :100:

**Comment by ajwerner** [comment]
TFTR!

bors r+

File: pkg/sql/pgwire/types.go [hunk 1]
```diff
@@ -313,7 +314,9 @@ func (b *writeBuffer) writeTextColumnarElement(
 		b.writeLengthPrefixedString(d.String())
 
 	case types.BytesFamily:
-		writeTextBytes(b, string(vecs.BytesCols[colIdx].Get(rowIdx)), conv)
+		writeTextBytes(
+			b, unsafeBytesToString(vecs.BytesCols[colIdx].Get(rowIdx)), conv,
```
Comment (rafiss) [inline pkg/sql/pgwire/types.go:318]: i don't really know what the contracts are with the `columnar` write thing. do we know that this is safe?

Comment (ajwerner) [reply pkg/sql/pgwire/types.go:318]: I'm almost certain, yes. The deal here is that we're subslicing a bigger slice, which is fine in that that slice is not changing during this call (if it were, we'd have a race). As far as I can tell, `writeTextBytes` does not retain the string after it returns; it appends its bytes to an underlying buffer.

File: pkg/sql/pgwire/types.go [hunk 2]
```diff
@@ -178,7 +179,7 @@ func writeTextDatumNotNull(
 		b.writeLengthPrefixedDatum(v)
 
 	case *tree.DBytes:
-		writeTextBytes(b, string(*v), conv)
+		writeTextBytes(b, unsafeBytesToString([]byte(*v)), conv)
```
Comment (jordanlewis) [inline pkg/sql/pgwire/types.go:182]: Hmm I feel like there is something weird going on here - a `DBytes` is a string already. Calling string(*v) where v is a pointer to DBytes shouldn't cause a copy? What is the expense here exactly?

Comment (ajwerner) [reply pkg/sql/pgwire/types.go:182]: Yeah, you're right, I just got overzealous and wasn't paying attention. Give me a sec.

**Review by jordanlewis — APPROVED** [review]

**Pull request merged**